### PR TITLE
Feature/blueprint config

### PR DESCRIPTION
**Pull request by shapirov103**
*Issue #, if available:* #435

*Description of changes:* General improvement of cluster and add-on configurability that allows running arbitrary functions in the stack context to provision resources that may be reused throughout the stack. Simple use cases, like passing custom security group or masters role to the cluster are simplified and enabled without extra stacks.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/cluster-providers/generic-cluster-provider.md [hunk 1]
```diff
@@ -16,7 +16,11 @@ Full list of configuration options:
 ```typescript
 const clusterProvider = new blueprints.GenericClusterProvider({
     version: KubernetesVersion.V1_23,
-    serviceIpv4Cidr: "10.43.0.0/16"
+    serviceIpv4Cidr: "10.43.0.0/16",
+    // if needed use this to register an auth role integrate with RBAC
```
Comment (elamaran11) [inline docs/cluster-providers/generic-cluster-provider.md:20]: to integrate with RBAC

File: test/resource-providers/resource-proxy.test.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        const template = Template.fromStack(stack1);
+        const templateString = JSON.stringify(template.toJSON());
+        expect(templateString).toContain(':iam::123456789012:role/myrole');
+        expect(templateString).toContain('ControlPlaneSG');
+        expect(templateString).toContain(sgDescription);
+        expect(nutil.isProxy(clusterProvider.props.securityGroup)).toBeTruthy();
+    });
+
+
+    // test("When a stack is created with proxy for launchTemplateSpec in cluster provider, stack is created with proxy resolved", () => {
```
Comment (elamaran11) [inline test/resource-providers/resource-proxy.test.ts:90]: Could we remove this commented code if its not needed.

File: lib/resource-providers/utils.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import { v4 as uuidv4 } from 'uuid';
+
+export function getNamedResource<T extends IResource = IResource>(resourceName : string) : T {
+    return new Proxy({} as T, new DummyProxy((resourceContext: ResourceContext) => {
+        return resourceContext.get(resourceName) as T;
+    }));
+}
+
+
+export function getResource<T extends IResource = IResource>(fn: (context: ResourceContext) => T) : T {
```
Comment (elamaran11) [inline lib/resource-providers/utils.ts:13]: Could we add couple of lines of documentation of what this function is doing?

File: lib/resource-providers/utils.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { IResource } from 'aws-cdk-lib/core';
+import { DummyProxy } from '../utils/proxy-utils';
+import { ResourceContext } from '../spi';
+import { v4 as uuidv4 } from 'uuid';
+
+export function getNamedResource<T extends IResource = IResource>(resourceName : string) : T {
```
Comment (elamaran11) [inline lib/resource-providers/utils.ts:6]: Could we add couple of lines of documentation of what this function is doing?

File: lib/utils/proxy-utils.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import * as nutil from 'node:util/types';
+
+export type OneArgFn<T> = (arg: any) => T;
+
+export const isDynamicProxy = Symbol("isDynamicProxy");
+export const sourceFunction = Symbol("sourceFunction");
+
+
```
Comment (elamaran11) [inline lib/utils/proxy-utils.ts:8]: Could we have documentation to these functions pls?

**Review by elamaran11 — COMMENTED** [review]
@shapirov103 Looks great. Just minor feedback for documentation and typos.

**Review by youngjeong46 — COMMENTED** [review]
Just one minor comment.

File: docs/cluster-providers/generic-cluster-provider.md (hunk 1 above)
Comment (shapirov103) [reply docs/cluster-providers/generic-cluster-provider.md:20]: addressed

File: test/resource-providers/resource-proxy.test.ts (hunk 2 above)
Comment (shapirov103) [reply test/resource-providers/resource-proxy.test.ts]: addressed

File: lib/resource-providers/utils.ts (hunk 3 above)
Comment (shapirov103) [reply lib/resource-providers/utils.ts:42]: addressed

File: lib/resource-providers/utils.ts (hunk 4 above)
Comment (shapirov103) [reply lib/resource-providers/utils.ts:21]: addressed

File: lib/utils/proxy-utils.ts (hunk 5 above)
Comment (shapirov103) [reply lib/utils/proxy-utils.ts:14]: addressed

File: docs/cluster-providers/generic-cluster-provider.md [hunk 6]
```diff
@@ -51,6 +55,44 @@ EksBlueprint.builder()
...
+    mastersRole: blueprints.getResource(context => {
+        return new iam.Role(context.scope, 'AdminRole', { assumedBy: new AccountRootPrincipal() });
+    }),
+    securityGroup: blueprints.getNamedResource("my-cluster-security-group"), // assumed to be register as a resource provider under name my-cluster-security-group
+    managedNodeGroups: [
+        {
+            id: "mng1",
+            nodeRole: blueprints.getResource(context => {
+                const role = new iam.Role(context.scope, 'NodeRole', { assumedBy: new new iam.ServicePrincipal("ec2.amazonaws.com")});
```
[the remark this answers is not in the document]
Comment (shapirov103) [reply docs/cluster-providers/generic-cluster-provider.md]: addressed

**Review by elamaran11 — APPROVED** [review]
LGTM

**Comment by shapirov103** [comment]
@youngjeong46 assuming you are good with the changes. If something comes up, I will re-open.

**Pull request merged**